src/contracts/signing-key-events-cache/signing-key-events-cache.service.ts

```diff
@@ -0,0 +1,395 @@
+
+    if (!isCacheValid) {
+      process.exit(1);
+    }
+
+    const wasUpdated = await this.stakingModuleListWasUpdated();
+    if (wasUpdated) {
+      this.logger.log('Staking module list was updated. Deleting cache');
+      await this.levelDBCacheService.deleteCache();
+    }
+
+    await this.updateEventsCache();
+  }
+
+  /**
+   * Updates the cache signing keys events
+   * The last N blocks are not stored, in order to avoid storing reorganized blocks
+   */
+  public async updateEventsCache(): Promise<number> {
+    const fetchTimeStart = performance.now();
+
+    const [latestBlock, initialCache] = await Promise.all([
+      this.providerService.getBlockNumber(),
+      this.getCachedEvents(),
+    ]);
+
+    const firstNotCachedBlock = initialCache.headers.endBlock + 1;
+    const toBlock = latestBlock - SIGNING_KEYS_EVENTS_CACHE_LAG_BLOCKS;
+
+    const totalEventsCount = initialCache.data.length;
+    let newEventsCount = 0;
+
+    for (
+      let block = firstNotCachedBlock;
+      block <= toBlock;
+      block += FETCHING_EVENTS_STEP
+    ) {
+      const chunkStartBlock = block;
+      const chunkToBlock = Math.min(toBlock, block + FETCHING_EVENTS_STEP - 1);
+
+      const chunkEventGroup = await this.fetchEventsFallOver(
+        chunkStartBlock,
+        chunkToBlock,
+      );
+
+      await this.levelDBCacheService.insertEventsCacheBatch({
+        headers: {
+          ...initialCache.headers,
+          // as we update staking modules addresses always before run of this method, we can update value on every iteration
+          stakingModulesAddresses: chunkEventGroup.stakingModulesAddresses,
+          endBlock: chunkEventGroup.endBlock,
+        },
+        data: chunkEventGroup.events,
+      });
+
+      newEventsCount += chunkEventGroup.events.length;
+
+      this.logger.log('Historical signing key add events are fetched', {
+        toBlock,
+        startBlock: chunkStartBlock,
+        endBlock: chunkToBlock,
+      });
+    }
+
+    const fetchTimeEnd = performance.now();
+    const fetchTime = Math.ceil(fetchTimeEnd - fetchTimeStart) / 1000;
+    // TODO: replace timer with metric
+
+    this.logger.log('Signing key events cache is updated', {
+      newEventsCount,
+      totalEventsCount: totalEventsCount + newEventsCount,
+      fetchTime,
+    });
+
+    return toBlock;
+  }
+
+  public async stakingModuleListWasUpdated(): Promise<boolean> {
+    const {
+      headers: { stakingModulesAddresses: previousModules },
+    } = await this.levelDBCacheService.getHeader();
+
+    if (!previousModules.length) {
+      return false;
+    }
+
+    const currentModules = await this.getStakingModules();
+    const isDifferentLength = currentModules.length !== previousModules.length;
+    const hasNewModules = currentModules.some(
+      (module) => !previousModules.includes(module),
+    );
+
+    const wasUpdated = isDifferentLength || hasNewModules;
```

Comment:
I see that it might not be obvious, and it is better to write it the way you did. However, I wrote a test to cover all cases, so you might find a case that it doesn't cover.

I will change the code as you suggested too